rfc-117-containerised-runtime.md

```diff
@@ -0,0 +1,69 @@
+
+A containerised runtime aligns with the [GOV.UK Infrastructure Architecture Goals for 2021](
+https://docs.google.com/document/d/1ooN7wkYhEGvceGe9Qz_HNZa-GPtrjzK_vA4vfWYVn4c/edit#heading=h.cdrr7rv9t98f) to isolate applications through containers. This will give us better control of how resources are matched to applications.
+
+The TechOps strategy is to use common components where available. GOV.UK has been following the strategy by using hosted versions of software ([ElasticSearch](https://aws.amazon.com/elasticsearch-service/) and [Postgres](https://aws.amazon.com/rds/)) and services ([Notify](https://www.notifications.service.gov.uk)) where available. Using a containerised runtime is a continuation of this policy.
+
+GOV.UK has already containerised it’s development environment ([rfc-106](https://www.github.com/alphagov/govuk-rfcs/106)), a containerised runtime will narrow the gap between development and production.
+
+### Benefits of this approach
+
+The most immediate benefit of this approach is that it will allow us to incrementally upgrade the operating system version. By moving the applications into containers, GOV.UK can enable a much lower risk upgrade path.
+
+Moving to a container based hosting environment will allow us to remove much of our legacy Puppet and Fabric code. This code is hard to maintain, and hard to hire people with the experience required to improve and upgrade. With industry and government overwhelmingly moving towards containerised platforms, it should be much easier to hire people with these skills.
+
+Containerised hosting environments can generally provide much better value for money in terms of resource usage, both because many small applications can be “bin-packed” onto the same infrastructure, and because it is much easier to autoscale applications.
+
+Running containers in production allows for a reduction in the difference between development environments and production. This can lead to bugs being found more quickly, and faster development cycles.
+
+Other improvements to GOV.UK’s development lifecycle (such as moving to continuous deployment) will be easier to implement on a container based infrastructure than they would in the current virtual machine based infrastructure.
+
+### Risks of this approach
+
+GOV.UK has not yet estimated the length of time this work will take, our current assumption is that it will take much less than 2 years to complete. There is a risk that the “unknown unknowns” GOV.UK discovers will prevent GOV.UK from finishing the containerisation work before support ends for Trusty. If this happens, GOV.UK may need to refocus our efforts to upgrade from Trusty in the existing infrastructure.
+
+GOV.UK infrastructure has been in a transitory state for several years and containerisation will prolong this state. In addition in the short term GOV.UK will be adding complexity making supporting the platform hard for RE-GOV.UK and 2ndline.
```

Comment:
Given this long timeframe, what are we doing and **what will we do to ensure the ongoing health of the _people_ who work on the teams doing this migration**? Is team structure going to change? Will there be an embedded infrastructure team again?

Reply:
This also feels like it will require investment of all devs - not just SREs - to do this properly, with consent, training, etc, and for devs to feel like they have agency in the future of GOV.UK's platform - because that's important because they're the ones supporting it.

Reply:
Building on @issyl0's point, we should also think about skills required during recruitment, as well as the potential need to upskill others within GDS should they switch to working on GOV.UK for whatever reason.

Reply:
Currently we have to upskill them in Puppet, whereas other parts of GDS are already container based (Verify, PaaS), so I'd argue that there's _less_ divergeance here.

Reply:
Yeah, big 👍  to this - the impact on people is a key risk here, and we should call it out. GitHub won't let me make a suggestion in this thread. I'll add one below.

Reply:
Richard's suggestion on wording is in https://github.com/alphagov/govuk-rfcs/pull/117#discussion_r386425034, for those who read this thread and don't want to scroll forever to find it.

Reply:
It's now part of the RFC 7327c7e4cce50ac8f11fd4de656c106d0934ce84